test/index.js

```diff
@@ -0,0 +1,359 @@
+'use strict';
+
+var assert = require('assert');
+
+var Z = require('..');
+var Identity = require('./Identity');
+
+
+var eq = function(actual, expected) {
+  assert.strictEqual(arguments.length, 2);
+  assert.strictEqual(Z.toString(actual), Z.toString(expected));
+};
+
+
+var Point = function Point(x, y) {
+  this.x = x;
+  this.y = y;
+};
+Point.prototype.z = 0;
+
+
+var method = function() {};
+
+//  identity :: a -> a
+var identity = function(x) { return x; };
+
+//  inc :: Number -> Number
+var inc = function(x) { return x + 1; };
+
+//  add :: (Number, Number) -> Number
+var add = function(x, y) { return x + y; };
+
+//  square :: Number -> Number
+var square = function(x) { return x * x; };
+
+//  duplicate :: a -> Pair a a
+var duplicate = function(x) { return [x, x]; };
+
+//  length :: List a -> Integer
+var length = function(xs) { return xs.length; };
+
+//  prependTo :: (Array a, a) -> Array a
+var prependTo = function(xs, x) { return [x].concat(xs); };
+
+//  _repeat :: Array a -> a -> Integer -> Array a
+var _repeat = function _repeat(xs) {
+  return function(x) {
+    return function(n) {
+      return n === 0 ? xs : _repeat(xs.concat([x]))(x)(n - 1);
+    };
+  };
+};
+
+//  repeat :: a -> Integer -> Array a
+var repeat = _repeat([]);
+
+//  numArgs :: (Any...) -> Integer
+var numArgs = function() { return arguments.length; };
+
+
+describe('TypeClass', function() {
+
+  it('is a ternary function', function() {
+    eq(typeof Z.TypeClass, 'function');
+    eq(Z.TypeClass.length, 3);
+  });
+
+  it('defines a type class', function() {
+    //  Ord :: TypeClass
+    var Ord = Z.TypeClass(
+      'my-package/Ord',
+      [],
+      function(x) {
+        switch (Object.prototype.toString.call(x)) {
+          case '[object Date]':
+          case '[object Number]':
+          case '[object String]':
+            return true;
+          default:
+            return false;
+        }
+      }
+    );
+    eq(Ord['@@type'], 'sanctuary-type-classes/TypeClass');
+    eq(Ord.name, 'my-package/Ord');
+    eq(Ord.test(42), true);
+    eq(Ord.test([]), false);
+    eq(Ord.toString(), 'Ord');
+  });
+
+});
+
+describe('type classes', function() {
+
+  it('Setoid.test', function() {
+    eq(Z.Setoid.test(null), false);
+    eq(Z.Setoid.test([]), true);
+    eq(Z.Setoid.test({}), true);
+    eq(Z.Setoid.test(''), true);
+  });
+
+  it('Semigroup.test', function() {
+    eq(Z.Semigroup.test(null), false);
+    eq(Z.Semigroup.test([]), true);
+    eq(Z.Semigroup.test({}), true);
+    eq(Z.Semigroup.test(''), true);
+  });
+
+  it('Monoid.test', function() {
+    eq(Z.Monoid.test(null), false);
+    eq(Z.Monoid.test([]), true);
+    eq(Z.Monoid.test({}), true);
+    eq(Z.Monoid.test(''), true);
+  });
+
+  it('Functor.test', function() {
+    eq(Z.Functor.test(null), false);
+    eq(Z.Functor.test([]), true);
+    eq(Z.Functor.test({}), true);
+    eq(Z.Functor.test(''), false);
+    eq(Z.Functor.test(inc), true);
```

Comment:
The problem is that having defined `Object#fantasy-land/map`, all plain JavaScript objects satisfy the predicate. I did add these assertions, though:

```javascript
eq(Z.Functor.test({'@@type': 'my-package/Foo', map: method}), false);
eq(Z.Functor.test({'@@type': 'my-package/Foo', 'fantasy-land/map': null}), false);
eq(Z.Functor.test({'@@type': 'my-package/Foo', 'fantasy-land/map': method}), true);
```

All the `<Type>.test` assertions were useful during development, but I believe they've outlived their usefulness. `eq(Z.Chain.test([]), true)` only asserts that `methods` contains an entry for `Array#fantasy-land/chain`. We could assert this indirectly by testing `Z.chain`.

What I'd like to do is write more tests for `Z.TypeClass` so we can have enough confidence in

```javascript
var Chain = Z.Chain =
TypeClass('sanctuary-type-classes/Chain',
          [Apply],
          hasMethod('fantasy-land/chain'));
```

that we're not obliged to test `Z.Chain.test` and all the other predicates individually.